extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/IOThreadClientInterceptor.java

```diff
@@ -21,21 +34,38 @@
             @Override
             public void start(Listener<RespT> responseListener, Metadata headers) {
                 super.start(new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
+                    private volatile CompletableFuture<Void> onMessageCompletion;
+
                     @Override
                     public void onMessage(RespT message) {
-                        runInContextIfNeed(() -> super.onMessage(message));
+                        if (context != null) {
+                            onMessageCompletion = new CompletableFuture<>();
+                            context.runOnContext(unused -> {
+                                try {
+                                    super.onMessage(message);
+                                    onMessageCompletion.complete(null);
+                                } catch (Throwable any) {
+                                    onMessageCompletion.completeExceptionally(any);
+                                }
+                            });
+                        } else {
+                            super.onMessage(message);
+                        }
                     }
 
                     @Override
                     public void onClose(Status status, Metadata trailers) {
-                        runInContextIfNeed(() -> super.onClose(status, trailers));
-                    }
-
-                    private void runInContextIfNeed(Runnable fun) {
-                        if (context != null) {
-                            context.runOnContext(unused -> fun.run());
+                        if (onMessageCompletion != null && !Context.isOnEventLoopThread()) {
+                            try {
+                                onMessageCompletion.get(60, TimeUnit.SECONDS);
+                            } catch (InterruptedException | ExecutionException e) {
+                                throw new RuntimeException("`onMessage` failed or interrupted", e);
+                            } catch (TimeoutException e) {
+                                throw new RuntimeException("`onMessage` did not complete in 60 seconds");
+                            }
+                            super.onClose(status, trailers);
                         } else {
-                            fun.run();
+                            super.onClose(status, trailers);
```

Comment:
shouldn't you enqueue close to the `context` if any?

Reply:
If I do that, the newly added test gets stuck. Do you know what problem was @Andro999b fixing by making the `onClose` executed on the event loop?

Reply:
Hi @michalszynkiewicz. There was problem that onClose and onMessage was executed on different threads. Becuse of this when you use grpc streams you may loss messages becuse onClose can be called before onMessage. You may found more details in this issue https://github.com/quarkusio/quarkus/issues/15187

Reply:
Thanks Andrii! I'll take your reproducer and see if with my changes it still works. And then bother Clement to figure out what's the proper solution :)

Reply:
@michalszynkiewicz reproducer that i have in the issue not really work. I uploaded correct one here: 
https://drive.google.com/file/d/16xdsSuSV4V9pQAJt_HDYoppgqsAGOGoX/view?usp=sharing

Reply:
Do I get correctly that we have an error if the test (executed multiple times) sometimes prints this?
```
java.util.NoSuchElementException
        at io.grpc.stub.ClientCalls$BlockingResponseStream.next(ClientCalls.java:662)
        at StreamClientTest$MyConsumer.lambda$consume$1(StreamClientTest.java:57)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
null
true

```


With my changes I don't see this error too.

Reply:
@michalszynkiewicz Yep, correct.